TWILL-63, combining twill classes into a twill.jar and having other u…

gsps1 opened the pull request:
…ser classes into a program.jar, also caching the twill.jar in YarnTwillRunnerService and the content is reused across TwillApplications started by the runner

chtyim commented on twill-core/src/main/java/org/apache/twill/internal/ApplicationBundler.java:
```diff
@@ -122,15 +122,16 @@ public boolean accept(String className, URL classUrl, URL classPathUrl) {
 
-  public void createBundle(Location target, Iterable<Class<?>> classes) throws IOException {
-    createBundle(target, classes, ImmutableList.<URI>of());
+  public void createBundle(Location target, Iterable<Class<?>> classes, boolean skipTwillClasses) throws IOException {
```
Why need to introduce this specific boolean? Can't we use the constructor that takes `ClassAcceptor`?

gsps1 replied:
yeah, i should use that.

chtyim commented on twill-core/src/main/java/org/apache/twill/internal/ApplicationBundler.java:
```diff
@@ -178,8 +181,42 @@ public void createBundle(Location target, Iterable<Class<?>> classes, Iterable<U
+   * @return ByteArrayOutputStream
+   * @throws IOException
+   */
+  public ByteArrayOutputStream getBundleAsStream(Iterable<Class<?>> classes,
```
So this could end up using a lot of heap space (and double it when `ByteArrayOutputStream.toByteArray()` is being called). Why it needs to be kept all in memory?

chtyim replied:
Also, why this method is in ApplicationBundler? Seems totally unrelated to what the role of the class is.

chtyim commented on twill-yarn/src/main/java/org/apache/twill/yarn/YarnTwillPreparer.java:
```diff
@@ -137,7 +139,7 @@
                     YarnAppClient yarnAppClient, ZKClient zkClient,
                     LocationFactory locationFactory, String extraOptions, LogEntry.Level logLevel,
-                    YarnTwillControllerFactory controllerFactory) {
+                    YarnTwillControllerFactory controllerFactory, byte[] twillJarContent) {
```
Shouldn't be `byte[]`. Better be a `InputSupplier<? extends InputStream>`.

gsps1 commented:
Thanks for the review @chtyim , i have addressed your comments, please review again, thanks!

chtyim commented on twill-yarn/src/main/java/org/apache/twill/yarn/YarnTwillRunnerService.java:
```diff
@@ -175,6 +187,36 @@ protected void shutDown() throws Exception {
     };
   }
 
+  public Location createTwillJar() throws IOException {
```
Why public?

chtyim commented on twill-yarn/src/main/java/org/apache/twill/yarn/YarnTwillRunnerService.java:
```diff
@@ -175,6 +187,36 @@ protected void shutDown() throws Exception {
+    classes.add(yarnAppClient.getClass());
+
+    twillDependencyClasses = getTwillDependencyClasses(classes);
+    File tempFile = File.createTempFile("twill", ".jar");
```
Potentially this file can be deleted by the system when the runner process runs for a long time.

chtyim commented on twill-yarn/src/main/java/org/apache/twill/yarn/YarnTwillRunnerService.java:
```diff
@@ -175,6 +187,36 @@ protected void shutDown() throws Exception {
+
+    twillDependencyClasses = getTwillDependencyClasses(classes);
+    File tempFile = File.createTempFile("twill", ".jar");
+    Location twillJar = new LocalLocationFactory().create(tempFile.toURI());
```
Why need to wrap it to location? Can't it be just `File`?

chtyim commented on twill-yarn/src/main/java/org/apache/twill/yarn/YarnTwillRunnerService.java:
```diff
@@ -175,6 +187,36 @@ protected void shutDown() throws Exception {
+    if (classLoader == null) {
+      classLoader = getClass().getClassLoader();
+    }
+    return Dependencies.getClassDependencies(classLoader, new ClassAcceptor(), classNames);
```
The `getClassDependencies` doesn't give you all twill classes. It only gives you all classes in the dependencies. I think the better way is using all classPathURL encountered during the creation of the twill jar to create a `URLClassLoader`. To check if a class is inside the twill.jar, you can use `ClassLoader.getResource()` on the `URLClassLoader`.

gsps1 commented:
@chtyim have addressed your comments, please review, thanks

chtyim commented on twill-common/src/main/java/org/apache/twill/internal/Constants.java:
```diff
@@ -51,6 +51,10 @@
 
+  /** Twill Jar location configuration name and the default value for that config */
+  public static final String TWILL_JAR_LOCATION = "twill.jar.location";
+  public static final String DEFAULT_TWILL_JAR_LOCATION = "/tmp/";
```
Default should be the system tmp, which may not be `/tmp`. You can get it through `System.getProperty("java.io.tmpdir")`

chtyim commented on twill-common/src/main/java/org/apache/twill/internal/Constants.java:
```diff
@@ -64,8 +68,8 @@
-    public static final String APP_MASTER_JAR = "appMaster.jar";
-    public static final String CONTAINER_JAR = "container.jar";
+    public static final String TWILL_JAR = "twill.jar";
+    public static final String PROGRAM_JAR = "program.jar";
```
Better call it `application.jar`.

chtyim commented on twill-core/src/main/java/org/apache/twill/internal/ApplicationBundler.java:
```diff
@@ -145,41 +145,63 @@ public void createBundle(Location target, Class<?> clz, Class<?>...classes) thro
+    createBundleAndGetClassPathUrls(target, classes, resources);
+  }
+
+  public Set<URL> createBundleAndGetClassPathUrls(Location target, Iterable<Class<?>> classes,
```
Who call this? Seems like it's only called in this class (from above method). Maybe the private one can be called directly?

chtyim commented on twill-core/src/main/java/org/apache/twill/internal/ApplicationBundler.java:
```diff
@@ -145,41 +145,63 @@ public void createBundle(Location target, Class<?> clz, Class<?>...classes) thro
+  public Set<URL> createBundleAndGetClassPathUrls(Location target, Iterable<Class<?>> classes,
+                                                  Iterable<URI> resources) throws IOException {
+    return createBundleAndGetClassPathUrls(target.getName(), target.toString(),
+                                           target.getOutputStream(), classes, resources);
```
Also, this leaks the output stream resource.

chtyim commented on twill-core/src/main/java/org/apache/twill/internal/ApplicationBundler.java:
```diff
@@ -145,41 +145,63 @@ public void createBundle(Location target, Class<?> clz, Class<?>...classes) thro
+  }
+
+  private Set<URL> createBundleAndGetClassPathUrls(String targetName, String targetPath,
+                                                   OutputStream targetOutputStream,
```
If it takes output stream, then why need to take name and path? You either take a `Location` or `File`, then create an output stream and manage and closing it in this method, or you take a OutputStream and write to it, without caring where does it actually output to.

chtyim commented on twill-common/src/main/java/org/apache/twill/internal/Constants.java:
```diff
@@ -51,6 +51,10 @@
   public static final String RESTART_RUNNABLES_INSTANCES = "restartRunnablesInstances";
 
+  /** Twill Jar location configuration name and the default value for that config */
+  public static final String TWILL_JAR_LOCATION = "twill.jar.location";
```
What is the location? Is it a local directory or a directory based on the `LocationFactory`? Also, better make it clear that it's a directory.

chtyim commented on twill-core/src/main/java/org/apache/twill/internal/ApplicationBundler.java:
```diff
@@ -145,41 +145,63 @@ public void createBundle(Location target, Class<?> clz, Class<?>...classes) thro
+    LOG.debug("start creating bundle {}. building a temporary file locally at first", targetName);
     // Write the jar to local tmp file first
-    File tmpJar = File.createTempFile(target.getName(), ".tmp");
+    File tmpJar = File.createTempFile(targetName, ".tmp");
```
So this will be using the system temp directory. Maybe better use the one that user configurated

chtyim commented on twill-core/src/main/java/org/apache/twill/internal/ApplicationBundler.java:
```diff
@@ -145,41 +145,63 @@ public void createBundle(Location target, Class<?> clz, Class<?>...classes) thro
       LOG.debug("cleaned up local temporary for bundle {}", tmpJar.toURI());
     }
+    return classPathUrls;
+
```
remove extra new line.

chtyim commented on twill-yarn/src/main/java/org/apache/twill/yarn/YarnTwillRunnerService.java:
```diff
@@ -175,6 +189,27 @@ protected void shutDown() throws Exception {
+      normalizeTwillURLs.add(new File(url.getPath()).toURI().toURL());
+    }
+
+    twillClassLoader = new URLClassLoader(normalizeTwillURLs.toArray(new URL[normalizeTwillURLs.size()]), null);
```
Who will close this ClassLoader?

chtyim replied:
Oh, seems like it will get reused. That's ok then.

chtyim commented on twill-core/src/main/java/org/apache/twill/internal/ApplicationBundler.java:
```diff
@@ -145,41 +145,63 @@ public void createBundle(Location target, Class<?> clz, Class<?>...classes) thro
+
+  }
+
+  public Set<URL> createBundleAndGetClassPathUrls(File target, Iterable<Class<?>> classes,
```
no doc.

hsaputra commented:
Any update on this PR?

gsps1 commented:
thanks for the review @chtyim , @hsaputra i will be addressing the comments soon, got busy with other things. will update soon. Thanks for following up!

hsaputra commented:
Could you kindly move the PR to the new Twill repo: https://github.com/apache/twill

Thanks
